### ISSUE TITLE
Major performance degradation when cache is enabled on 0.0.274

### ISSUE THREAD

**Issue by denik** [astral-sh/ruff#5269]
When cache is full, ruff 0.0.274 is much slower than without cache:

```
% ruff --version
ruff 0.0.274

% rm -fr .ruff_cache

% time ruff -ns .
ruff -ns .  10.87s user 2.25s system 620% cpu 2.114 total

% time ruff -s .
ruff -s .  10.96s user 2.53s system 705% cpu 1.913 total

% time ruff -s .
ruff -s .  23.84s user 8.42s system 105% cpu 30.517 total

% time ruff -s .
ruff -s .  23.19s user 8.65s system 105% cpu 30.123 total
```

Previous version works as expected - full cache speeds it up:

```
% ruff --version
ruff 0.0.273

% rm -fr .ruff_cache

% time ruff -ns .
ruff -ns .  10.88s user 2.29s system 766% cpu 1.717 total

% time ruff -s .
ruff -s .  11.06s user 3.03s system 706% cpu 1.993 total

% time ruff -s .
ruff -s .  0.69s user 1.61s system 218% cpu 1.053 total

% time ruff -s .
ruff -s .  0.69s user 1.60s system 219% cpu 1.044 total
```

```
% git ls-files '*.py' | wc -l
   38541
```

This is on Mac OS X, but the same effect is observed on Linux.

**Comment by charliermarsh** [comment]
Thanks, that's a surprisingly extreme degradation, will take a look.

Can you tell me a bit about the characteristics of the project? That's an enormous number of Python files :) Do most of the files contain violations? Or is `ruff check` clean? (Is this a public repo?)

**Comment by charliermarsh** [comment]
I would love to test `main` on this to see if it's at all improved.

**Comment by denik** [comment]
This was a private repo with a huge number of violations. 

```
Found 406804 errors.
[*] 3152 potentially fixable with the --fix option.
```

**Comment by charliermarsh** [comment]
In my testing, on large projects like CPython, `main` is faster or as-fast as v0.0.273, while v0.0.274 is 5-10x slower. So I'd expect the next release to fix this to a large degree, but the degradation you're seeing is even larger than that 5-10x estimate, so it's hard to say with complete certainty.

**Comment by charliermarsh** [comment]
Maybe I'll cut another release sooner rather than later and we can test it out. I'm trying to find another project of comparable scale. CPython is the largest I can find, at 696,787 lines of Python across 1,893 files.

**Comment by charliermarsh** [comment]
Okay, I tested on TensorFlow, which has 925,816 lines of Python across 3,059 files, and `main` and `v0.0.273` are equivalent, while `v0.0.274` is 15x slower.

**Comment by denik** [comment]
Checked e71f044f0d6b6869 - it's indeed much faster, similar to 0.0.273.

Thanks! Looking forward to 0.0.275 :)

**Comment by charliermarsh** [comment]
Oh, thank you so much for doing that! That's really helpful to hear. I probably won't release tonight (just a little drained from yesterday's release cycle), but before the end of the week.

**Comment by charliermarsh** [comment]
Should be fixed in v0.0.275 (out now).
